src/hooks/useGridWidth.ts

```diff
@@ -1,23 +1,65 @@
 import { useRef, useState, useLayoutEffect } from 'react';
 
+// https://github.com/microsoft/TypeScript/issues/37861
+interface ResizeObserverSize {
+  inlineSize: number;
+}
+
+interface ResizeObserverEntry {
+  target: Element;
+  borderBoxSize?: ResizeObserverSize;
+}
+
+type ResizeObserverCallback = (entries?: ResizeObserverEntry[]) => void;
+
+type ResizeObserver = new (callback: ResizeObserverCallback) => {
+  observe: (target: Element) => void;
+  unobserve: (target: Element) => void;
+};
+
+
+const { ResizeObserver } = window as (typeof window & { ResizeObserver: ResizeObserver });
+
 export function useGridWidth<T>(width?: number): [React.RefObject<HTMLDivElement>, number] {
   const gridRef = useRef<HTMLDivElement>(null);
+  const [resizeObserver] = useState(() => {
+    if (ResizeObserver === undefined) {
+      return null;
+    }
+
+    return new ResizeObserver((entries?: ResizeObserverEntry[]) => {
+      if (Array.isArray(entries) && entries.length > 0) {
+        const newWidth = entries[0].borderBoxSize?.inlineSize ?? entries[0].target.getBoundingClientRect().width;
```

Comment:
`borderBoxSize` is not supported everywhere and it is marked as experimental
https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserverEntry/borderBoxSize